src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,263 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+
+import "../interfaces/IIndexRegistry.sol";
+import "../interfaces/IRegistryCoordinator.sol";
+import "../libraries/BN254.sol";
+
+contract IndexRegistry is IIndexRegistry {
+
+    /// @notice The value that indices of deregistered operators are set to
+    uint32 public constant OPERATOR_DEREGISTERED_INDEX = type(uint32).max;
+
+    IRegistryCoordinator public immutable registryCoordinator;
+
+    // list of all operators ever registered, may include duplicates. used to avoid running an indexer on nodes
+    bytes32[] public globalOperatorList;
+
+    // mapping of operatorId => quorumNumber => index history of that operator
+    mapping(bytes32 => mapping(uint8 => OperatorIndexUpdate[])) internal _operatorIdToIndexHistory;
+    // mapping of quorumNumber => history of numbers of unique registered operators
+    mapping(uint8 => OperatorIndexUpdate[]) internal _totalOperatorsHistory;
+
+    modifier onlyRegistryCoordinator() {
+        require(msg.sender == address(registryCoordinator), "IndexRegistry.onlyRegistryCoordinator: caller is not the registry coordinator");
+        _;
+    }
+
+    constructor(
+        IRegistryCoordinator _registryCoordinator
+    ){
+        registryCoordinator = _registryCoordinator;
+    }
+
+    /**
+     * @notice Registers the operator with the specified `operatorId` for the quorums specified by `quorumNumbers`.
+     * @param operatorId is the id of the operator that is being registered
+     * @param quorumNumbers is the quorum numbers the operator is registered for
+     * @return numOperatorsPerQuorum is a list of the number of operators (including the registering operator) in each of the quorums the operator is registered for
+     * @dev access restricted to the RegistryCoordinator
+     * @dev Preconditions (these are assumed, not validated in this contract):
+     *         1) `quorumNumbers` has no duplicates
+     *         2) `quorumNumbers.length` != 0
+     *         3) `quorumNumbers` is ordered in ascending order
+     *         4) the operator is not already registered
+     */
+    function registerOperator(bytes32 operatorId, bytes calldata quorumNumbers) external onlyRegistryCoordinator returns(uint32[] memory) {
+        uint32[] memory numOperatorsPerQuorum = new uint32[](quorumNumbers.length);
+        //add operator to operatorList
+        globalOperatorList.push(operatorId);
```

Comment:
pointing out that we will push an operator repeatedly to this list when they register for new quorums. seems fine though, given we've already said the list 'may include duplicates' ?

Reply:
talked offline, this is technically acceptable.